cmd/gvisor-website/main.go

```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
 	})))
 }
 
+// registerRepo registers the repository handler.
+func registerRepo(mux *http.ServeMux) {
+	if mux == nil {
+		mux = http.DefaultServeMux
+	}
+
+	mux.Handle("/gvisor/git-upload-pack", hostRedirectHandler(http.HandlerFunc(gitUploadPack)))
+	mux.Handle("/gvisor/info/refs", hostRedirectHandler(http.HandlerFunc(infoRefs)))
+}
+
+func gitUploadPack(w http.ResponseWriter, r *http.Request) {
+	target, err := url.Parse(upstreamGitUploadPack)
+	if err != nil {
+		panic(err) // Should never happen.
+	}
+	// Proxy to the upstream repository.
+	proxy := &httputil.ReverseProxy{
+		Director: func(req *http.Request) {
+			req.URL.Scheme = target.Scheme
+			req.URL.Host = target.Host
+			req.URL.Path = target.Path
+			req.Header.Set("Host", target.Host)
+		},
+	}
+	proxy.ServeHTTP(w, r)
+}
+
+func infoRefs(w http.ResponseWriter, r *http.Request) {
+	// We intercept the client request. We implement only the appropriate
+	// git-upload-pack service and enforce that clients are asking for this
+	// as well. The full specification of this service, as well as a formal
+	// grammar is available at:
+	//
+	//	https://gist.github.com/schacon/6092633#file-git-http-proto-txt-L224
+	//
+	// We read only the first reference and change the primary branch. All
+	// other references and branches are left alone.
+
+	// Was this a dumb client? These are not supported.
+	vs, ok := r.URL.Query()["service"]
+	if !ok || len(r.URL.Query()) > 1 || len(vs) != 1 || vs[0] != "git-upload-pack" {
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprintf(w, "invalid query: %v", r.URL.Query())
+		return
+	}
+
+	// Connect upstream.
+	client := new(http.Client)
+	resp, err := client.Get(upstreamInfoRefs)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "upstream get error: %v", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	// Read the full upstream contents.
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "upstream read error: %v", err)
+		return
+	}
+
+	// Was there an error? Pass all data through.
+	if resp.StatusCode != 200 {
+		w.WriteHeader(resp.StatusCode)
+		w.Write(data)
+		return
+	}
+
+	// emitPkt emits a single packet line.
+	emitPkt := func(m string) {
+		if len(m) == 0 {
+			// Special case: normally the size includes the size of
+			// the four byte header. However, we see that size 0000
+			// appears after the header and is the terminal. We use
+			// the empty string to indicate this.
+			fmt.Fprintf(w, "%04x", 0)
+		} else {
+			fmt.Fprintf(w, "%04x%s\n", 4+len(m)+1, m)
+		}
+	}
+
+	// readPkt reads a single packet line.
+	readPkt := func() (string, bool) {
+		// Parse the size header and return the string. Note that we
+		// don't do bounds checking here and rely on the recovery
```

Comment:
Is this true? It looks like there is bounds checking going on and I don't see a recover.

Reply:
This was a previous iteration and the comment was stale. I decided to make it cleaner without that. Remove obsolete comment.